File: MIDTERM-LAB-3/MP04_CountValidRows.js
/**
 * =====================================================
 * Student Name    : Alidio, Jacob A.
 * Course          : Programming 2
 * Machine Problem : MP04 – Count Valid Dataset Rows
 * School          : University of Perpetual Help System DALTA, Molino Campus
 * Date            : April 8, 2026
 * Runtime         : Node.js  (run with: node MP04_CountValidRows.js)
 *
 * Description:
 *   Prompts the user for a CSV file path using Node.js readline,
 *   reads the file synchronously with the fs module, then counts
 *   how many data rows are "valid" (every field is non-empty).
 *   The header row is excluded from counting. A summary showing
 *   total, valid, and invalid row counts is printed at the end.
 * =====================================================
 */

// ── SECTION 1: Module Imports ────────────────────────────────────────
const fs       = require('fs');       // File System module for reading the CSV
const readline = require('readline'); // readline creates an interactive prompt

// ── SECTION 2: User Input ────────────────────────────────────────────
// rl is the readline interface that connects to the terminal (stdin/stdout)
const rl = readline.createInterface({
    input : process.stdin,
    output: process.stdout
});

const line = '='.repeat(52);

console.log(line);
console.log('  MP04 – Count Valid Dataset Rows');
console.log(line);

// Ask the user to type the CSV file path, then process their answer
rl.question('  Enter CSV file path: ', (filePath) => {
    rl.close();                        // close the prompt after one answer
    countValidRows(filePath.trim());   // hand off to the processing function
});

// ── SECTION 3: Core Processing Function ─────────────────────────────
/**
 * Reads the CSV file at filePath, skips the header, and counts
 * valid rows (all fields non-empty) vs invalid rows (has empty field).
 *
 * @param {string} filePath  - absolute or relative path to the CSV file
 */
function countValidRows(filePath) {

    // ── Step 1: Read the file ──
    // Wrap in try/catch so a bad path shows a friendly error instead of crashing
    let rawContent;
    try {
        // readFileSync returns the full file as a string (utf8 encoding)
        rawContent = fs.readFileSync(filePath, 'utf8');
    } catch (err) {
        console.log('[ERROR] Could not read file: ' + err.message);
        return;
    }

    // ── Step 2: Split into lines ──
    // Split on newline; filter removes any completely blank lines
    const lines = rawContent
        .split('\n')
        .map(l => l.trim())           // trim carriage-returns / whitespace
        .filter(l => l.length > 0);  // discard empty lines

    if (lines.length === 0) {
        console.log('[ERROR] File is empty.');
        return;
    }

    // ── Step 3: Parse the header ──
    // The first non-blank line is treated as the header row
    const headerColumns = lines[0].split(',');   // array of column name strings

    // ── Step 4: Count valid and invalid data rows ──
    let totalRows   = 0;  // all data rows (excludes header)
    let validRows   = 0;  // rows where every field has content
    let invalidRows = 0;  // rows with at least one empty field

    // Start from index 1 to skip the header line
    for (let i = 1; i < lines.length; i++) {

        totalRows++;

        // fields is an array of the comma-separated values for this row
        // -1 equivalent: split with a limit keeps trailing empty strings
        const fields = lines[i].split(',');

        // Check every field; rowIsValid flips to false on first empty field
        const rowIsValid = fields.every(field => field.trim().length > 0);

        if (rowIsValid) {
            validRows++;
        } else {
            invalidRows++;
        }
    }

    // ── SECTION 4: Output ────────────────────────────────────────────
    console.log();
    console.log(line);
    console.log('  RESULTS');
    console.log(line);
    console.log(`  Columns detected : ${headerColumns.length}`);
    console.log(`  Total data rows  : ${totalRows}`);
    console.log(`  Valid rows       : ${validRows}`);
    console.log(`  Invalid rows     : ${invalidRows}`);
    console.log(line);
}

File: MIDTERM-LAB-3/MP04_CountValidRows.java
/**
 * =====================================================
 * Student Name    : Alidio, Jacob A.
 * Course          : Programming 2
 * Machine Problem : MP04 – Count Valid Dataset Rows
 * School          : University of Perpetual Help System DALTA, Molino Campus
 * Date            : April 8, 2026
 * Runtime         : Java (compile: javac MP04_CountValidRows.java)
 *                           (run:     java  MP04_CountValidRows)
 *
 * Description:
 *   Prompts the user for a CSV file path, reads every line with
 *   BufferedReader, and counts how many DATA rows are "valid" –
 *   meaning no field in that row is blank/empty. The header line
 *   is skipped and never counted. Totals for valid rows, invalid
 *   rows, and all data rows are printed as a formatted summary.
 * =====================================================
 */

import java.io.*;
import java.util.*;

/**
 * MP04 - Count Valid Dataset Rows
 * Dataset: Pearson VUE Exam Results - University of Perpetual Help System Molino
 *
 * Program Logic:
 * Reads the Pearson VUE CSV, skips the preamble, and validates every data row
 * against the 8 meaningful display columns. A row is valid when none of those
 * columns are empty. A two-pass approach is used: the first pass collects all
 * data rows and detects which columns are actually used; the second pass validates
 * and categorizes each row. Invalid rows are listed in a table with [EMPTY]
 * markers, and all long values are truncated for clean terminal alignment.
 */
public class MP04_CountValidRows {

    static final String[] COL_LABELS = {
        "Candidate", "Student/ Faculty/ NTE", "Exam",
        "Language", "Exam Date", "Score", "Result", "Time Used"
    };
    static final int[] COL_MAX = { 20, 10, 30, 8, 10, 5, 6, 13 };

    static String[] parseCSVLine(String line) {
        List<String> fields = new ArrayList<>();
        StringBuilder cur = new StringBuilder();
        boolean inQ = false;
        for (char c : line.toCharArray()) {
            if (c == '"')              { inQ = !inQ; }
            else if (c == ',' && !inQ) { fields.add(cur.toString().trim()); cur.setLength(0); }
            else                       { cur.append(c); }
        }
        fields.add(cur.toString().trim());
        return fields.toArray(new String[0]);
    }

    static String truncate(String str, int maxLen) {
        if (str == null) str = "";
        return str.length() > maxLen ? str.substring(0, maxLen - 1) + "\u2026" : str;
    }

    /** isValidRow - Returns true if all display columns have a non-empty value. */
    static boolean isValidRow(String[] row, int[] colIdx) {
        for (int d = 0; d < colIdx.length; d++) {
            int ci  = colIdx[d];
            String v = (ci >= 0 && ci < row.length) ? row[ci] : "";
            if (v.trim().isEmpty()) return false;
        }
        return true;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.println("============================================================");
        System.out.println("     MP04 - Count Valid Dataset Rows                        ");
        System.out.println("     Pearson VUE Exam Results                               ");
        System.out.println("============================================================");
        System.out.print("Enter the CSV dataset file path: ");
        String filePath = scanner.nextLine().trim();

        String[] csvHeader = null;
        int[]    colIdx    = new int[COL_LABELS.length];
        Arrays.fill(colIdx, -1);

        ArrayList<String[]> dataBuffer  = new ArrayList<>();
        ArrayList<String[]> invalidRows = new ArrayList<>();
        ArrayList<Integer>  invalidNums = new ArrayList<>();
        int validCount = 0, invalidCount = 0;

        try {
            BufferedReader reader = new BufferedReader(
                new InputStreamReader(new FileInputStream(filePath), "UTF-8"));
            String line;

            // ── Pass 1: collect header + data rows ────────────────────────
            while ((line = reader.readLine()) != null) {
                line = line.replace("\uFEFF", "").replace("\r", "");
                if (line.trim().isEmpty()) continue;
                String[] fields = parseCSVLine(line);

                if (csvHeader == null) {
                    if (fields.length > 0 && fields[0].equalsIgnoreCase("Candidate")) {
                        csvHeader = fields;
                        for (int d = 0; d < COL_LABELS.length; d++)
                            for (int c = 0; c < csvHeader.length; c++)
                                if (csvHeader[c].trim().equalsIgnoreCase(COL_LABELS[d])) { colIdx[d] = c; break; }
                    }
                    continue;
                }
                if (fields[0].trim().isEmpty()) continue;
                dataBuffer.add(fields);
            }
            reader.close();

            if (csvHeader == null) { System.out.println("Error: Header row not found."); return; }

            // ── Pass 2: validate rows ─────────────────────────────────────
            int rowNum = 0;
            for (String[] fields : dataBuffer) {
                rowNum++;
                if (isValidRow(fields, colIdx)) {
                    validCount++;
                } else {
                    invalidCount++;
                    invalidRows.add(fields);
                    invalidNums.add(rowNum);
                }
            }

            int totalRows = dataBuffer.size();
            double pct = (totalRows > 0) ? (validCount * 100.0 / totalRows) : 0;

            // ── Summary ───────────────────────────────────────────────────
            System.out.println("\nFile   : " + filePath);
            System.out.println("\u2500".repeat(50));
            System.out.printf("  %-32s %d%n",     "Total columns (checked):", COL_LABELS.length);
            System.out.printf("  %-32s %d%n",     "Total data rows read:",    totalRows);
            System.out.printf("  %-32s %d%n",     "Valid rows:",              validCount);
            System.out.printf("  %-32s %d%n",     "Invalid (incomplete) rows:", invalidCount);
            System.out.printf("  %-32s %.2f%%%n", "Data completeness:",       pct);
            System.out.println("\u2500".repeat(50));

            if (invalidCount == 0) {
                System.out.println("\nAll rows are valid. No missing fields detected."); return;
            }

            // ── Invalid row detail table ──────────────────────────────────
            System.out.println("\nInvalid Row Details ([EMPTY] = missing field):");

            int numCols = COL_LABELS.length;
            int[] colW  = new int[numCols];
            for (int d = 0; d < numCols; d++) colW[d] = Math.max(COL_LABELS[d].length(), 7);

            String[][] displayRows = new String[invalidRows.size()][numCols];
            for (int r = 0; r < invalidRows.size(); r++) {
                String[] row = invalidRows.get(r);
                for (int d = 0; d < numCols; d++) {
                    int ci  = colIdx[d];
                    String raw = (ci >= 0 && ci < row.length) ? row[ci] : "";
                    String val = raw.trim().isEmpty() ? "[EMPTY]" : truncate(raw, COL_MAX[d]);
                    displayRows[r][d] = val;
                    colW[d] = Math.max(colW[d], val.length());
                }
            }

            int rowNumW = Math.max(4, String.valueOf(totalRows).length());
            StringBuilder sb = new StringBuilder("+-" + "-".repeat(rowNumW) + "-+");
            for (int w : colW) sb.append("-".repeat(w + 2)).append("+");
            String sep = sb.toString();

            System.out.println("\n" + sep);
            System.out.printf("| %-" + rowNumW + "s |", "Row#");
            for (int d = 0; d < numCols; d++) System.out.printf(" %-" + colW[d] + "s |", COL_LABELS[d]);
            System.out.println();
            System.out.println(sep);

            for (int r = 0; r < displayRows.length; r++) {
                System.out.printf("| %-" + rowNumW + "d |", invalidNums.get(r));
                for (int d = 0; d < numCols; d++) System.out.printf(" %-" + colW[d] + "s |", displayRows[r][d]);
                System.out.println();
            }
            System.out.println(sep);

        } catch (FileNotFoundException e) { System.out.println("Error: File not found -> " + filePath);
        } catch (IOException e)           { System.out.println("Error reading file: " + e.getMessage());
        } finally                         { scanner.close(); }
    }
}
